### ISSUE TITLE
Build mods don't work properly

### ISSUE THREAD

**Issue by jacovandermolen** [trevorsandy/lpub3d#520]
### Subject
Build mods don't work properly

### Environment
* Version of LPub3D 2.4.2
* Windows 10, 64 bit

### Configuration
Portable distro

### Steps to reproduce
Simple build mod test, trying to learn to do it in the 3D viewer.

```
0 Name: bm.ldr
0 Author: Jaco van der Molen [Jaco]
0 !LICENSE Not redistributable : see NonCAreadme.txt
0 !LPUB BUILD_MOD_ENABLED GLOBAL TRUE
1 1 0 0 0 0 -1 0 1 0 0 0 0 1 32523.dat
0 !LPUB BUILD_MOD BEGIN "bm.ldr Mod 1"
1 0 55 0 -20 1 0 0 0 0 -1 0 1 0 2780.dat
1 4 5 0 -20 0 0 1 1 0 0 0 1 0 HashL2.dat
0 !LPUB BUILD_MOD END_MOD
1 0 10 0 -20 1 0 0 0 0 -1 0 1 0 2780.dat
0 !LPUB BUILD_MOD END
0 STEP
0 !LPUB BUILD_MOD REMOVE "bm.ldr Mod 1"
1 14 20 0 -40 0 -1 0 1 0 0 0 0 1 32523.dat
0 STEP
```

Side note: moving a part in the X-direction using the 3D viewer also moves it a bit in the Z-direction. That should not happen?

Now not using a multi-step page, on page 2 the build mod is not removed and the pin is rendered at the wrong place.

![afbeelding](https://user-images.githubusercontent.com/32261647/113397162-e587de00-939c-11eb-9f3e-5676816184dc.png)
![afbeelding](https://user-images.githubusercontent.com/32261647/113397177-ecaeec00-939c-11eb-8fa1-130ef85736de.png)

### Expected behavior
On page 2 the build mod should be removed and the pin should be at the correct place.

### Actual behavior
Build mod is not applied.

### Workaround
Use a multistep page and refresh.
Or reload/refresh page 1 by doing so, the build mod is correct on page 2.

### Solution suggestion
Build mods should be applied on single-step pages.

**Comment by trevorsandy** [comment]
Thank you for reporting this behaviour. 

I added support for submodels within the build mod BEGIN command (your request) and  submodel preview which exposed a build mod design flaw. Correcting the flaw must have introduced this behaviour. I’ll take a look.

Cheers,

**Comment by trevorsandy** [comment]
I cannot reproduce this behaviour. I've tried on a fresh portable x86_64 distro of v2.4.2 r0 and on my latest DevOps instances.

Here's what I'm getting with the portable distro:

https://user-images.githubusercontent.com/13388970/113458832-f0ca2080-9413-11eb-8290-2072a5eeba19.mp4

I did notice on initial startup that it is necessary to set 'Archive Parts On Launch' and restart to load custom parts (i.e. helper parts). I don't like this flow so I'll add an action to execute parts archive on demand.

Cheers,

**Comment by trevorsandy** [comment]
Please try the latest continuous release. I will close this ticket. If you continue to experience the reported behaviour, please reopen.

Cheers,

**Comment by jacovandermolen** [comment]
OK, opening the file with build mods in place in 2.4.2 does work. I will let you know if a file without build mods and while working on it in LPub3D gives any trouble.

**Comment by jacovandermolen** [comment]
OK, Build mods are getting better.
However setting things up while working on a file is really difficult. Doing it in the 3D Viewer is not very intuative.
I'd love to share some thoughts on that, or see a demo from you how to make a build mod in LPub3D and the 3D Viewer.

I am getting the hang if it and got very good and expected results.

However, a fresh render (no cache, LPub folder, etc.) of a file that has Buildmods results in some strange rendering going through the pages. See this clip:
https://drive.google.com/file/d/1SODmZvyyy53sndv5GgLElP9U4O8tef1M/view?usp=sharing

Reaching page 4 renders some strange assemblies. Going to page 5 and going back to page 4 the expected result is rendered.
Same goes for going to page 6. Strange result, go to page 7 and return to page 6 is good.

Attached is the LDraw file I am using.
[40252.mpd.txt](https://github.com/trevorsandy/lpub3d/files/6267142/40252.mpd.txt)

**Comment by trevorsandy** [comment]
Thank you for your feedback. 

> However setting things up while working on a file is really difficult. Doing it in the 3D Viewer is not very intuative.

The approach I envision is to make the Build Mod user interaction natural. This is to say, the user navigates to a specific step which will be displayed in the 3DViewer/Editor. The user can then manipulate, in the 3DViewer/Editor, the step’s assembly using CUD. After which, the changes can be saved to a build modification for the specified step using the appropriate BuildMod menu action.

>  I'd love to share some thoughts on that, or see a demo from you how to make a build mod in LPub3D and the 3D Viewer.

I’ll create a clip that shows the nominal flow editing Build Mods from the 3DViewer.

Cheers,

**Comment by jacovandermolen** [comment]
I am applying more buildmods to the model and more steps are not good or vary from time to time going through the model by navigating the pages.
Here is the file I got so far, trying to remake the official instructions for set 40252 as close to the original as possible.
[40252.mpd.txt](https://github.com/trevorsandy/lpub3d/files/6273839/40252.mpd.txt)
The messed up renders (https://github.com/trevorsandy/lpub3d/issues/518) showed up again too now and then.

**Comment by jacovandermolen** [comment]
Currently testing with this model in 2.4.2.22.2684
[40252LPub42x.mpd.txt](https://github.com/trevorsandy/lpub3d/files/6276650/40252LPub42x.mpd.txt)
Here is a clip of what is happening:
https://drive.google.com/file/d/1_BSdc66mnKhtUWY6UFVHkH7u_aQeINZz/view?usp=sharing

**Comment by trevorsandy** [comment]
Corrected. 

3 items were treated:

1. Correct multi-step page with submodel displays invalid assembly

![Screenshot - 18_04_2021 , 14_48_05](https://user-images.githubusercontent.com/13388970/115146063-4974fc80-a055-11eb-92e5-bda19179b670.png)

![Screenshot - 18_04_2021 , 14_48_29](https://user-images.githubusercontent.com/13388970/115146101-68738e80-a055-11eb-8ad5-5829f30ff5e1.png)

![Screenshot - 18_04_2021 , 14_49_16](https://user-images.githubusercontent.com/13388970/115146141-893be400-a055-11eb-8625-cab70c89569e.png)

2. Correct line number when step bottom is new submodel

3. Correct (hopefully) page count for multi-step page with submodel

Cheers,

**Comment by jacovandermolen** [comment]
Great. Can I test in DevOps 2707 r45?

**Comment by trevorsandy** [comment]
Yes. This correction is stable. 

I'll be checking in some more updates in about 1 hour.

Cheers,

**Comment by trevorsandy** [comment]
I found more issues so the last DevOps build is not stable.

Cheers,

**Comment by jacovandermolen** [comment]
Is see. I have tested with my model, and all goes well (better) until the submodel that is called out. In the assembly the buildmod is not removed.

**Comment by trevorsandy** [comment]
> I found more issues so the last DevOps build is not stable.

Yep - that's the little bugger that I didn't catch.

Cheers,

**Comment by trevorsandy** [comment]
I just checked in the latest update. It should do the trick for the last outstanding build modification item.

There's are two items remaining to treat before I publish 2.4.3 - page count and command editor.

Cheers,

**Comment by jacovandermolen** [comment]
Great. I will test asap.

**Comment by jacovandermolen** [comment]
I think we are almost there. I did some tests with the set 40252 model. Most things look good and work as expected. In some cases, assemblies were rendered incorrect, and build mods are not properly handled. I will make a screen capture for my findings.
Displaying a render of a subassembly at the beginning of a submodel is rendered incorrect.
I'll show that too.

You have been quite busy debugging of course, but do you have an update on how to create build mods with the 3D editor?
I am still struggeling with that.

**Comment by trevorsandy** [comment]
Thank you for providing this feedback.

> In some cases, assemblies were rendered incorrect, and build mods are not properly handled.

Interesting. I look forward to seeing these cases.

> Displaying a render of a subassembly at the beginning of a submodel is rendered incorrect.

All Build Mods should be purged from the preview submodel. What aspects of the submodel are incorrectly rendered ?

Cheers,

**Comment by jacovandermolen** [comment]
I will post a video of my findings soon.

For now, there is one thing you might like to know is this:
Excerpt from my test file

1 321 0 0 0 1 0 0 0 1 0 0 0 1 3710.dat
0 !LPUB BUILD_MOD BEGIN "submodel-6.ldr Mod 2"
1 71 0 -44 0 1 0 0 0 1 0 0 0 1 3023.dat
0 !LPUB BUILD_MOD END_MOD
1 71 0 -8 0 1 0 0 0 1 0 0 0 1 3023.dat
0 !LPUB BUILD_MOD END
0 STEP
0 !LPUB BUILD_MOD REMOVE "submodel-06.ldr Mod 2"
1 71 -30 0 0 0 0 1 0 1 0 -1 0 0 50746.dat

I made a typo there in the REMOVE build mod 06 instead of 6.
On encountering that, LPub crashes with a dump file.
Perhaps catch that in an error or warning?

**Comment by trevorsandy** [comment]
> I made a typo there in the REMOVE build mod 06 instead of 6.
On encountering that, LPub crashes with a dump file.

I saw that when testing your example file but instead of an abend, I got a parse message as expected.

Perhaps after the last update the parse message is not triggered - I'll take a look.

Cheers,

**Comment by jacovandermolen** [comment]
Here is a recording with some findings
https://drive.google.com/file/d/1R-yDWAiXPLRdOo-bvt43oCr8tZIkd6d_/view?usp=drivesdk

This is the file 
[40252LPub42x.mpd.txt](https://github.com/trevorsandy/lpub3d/files/6356823/40252LPub42x.mpd.txt)

Clearing al cache and folders manually solves the false rendering of the 2x3 plate at step 14 and on (1:03") in the clip.

You have to go through the whole file page by page for everything to work properly.
If you directly navigate to say page 10, and go back and forth through the pages things start to mess up.

Redrawing the whole file or redraw a certain page or assembly messes things up too.

**Comment by jacovandermolen** [comment]
I also miss the option to make the border of a submodel preview with round corners.
And as you can see in the clip it is now rendered on every page for that submodel.

One other thing came up: the weird rendering I mentioned earlier came up again and I think this has to do with having the Visual Editor open and active. If not, things render just fine.

**Comment by trevorsandy** [comment]
Thank you for the recording.

> I also miss the option to make the border of a submodel preview with round corners.

Are you saying that when you updated the submodel preview (to display round corners), the submodel was displayed on all the following steps - versus just the first step as expectred ?

> the weird rendering I mentioned earlier came up again...

Are you talking about the rendering with artifacts #518 ?

Cheers,

**Comment by jacovandermolen** [comment]
> > I also miss the option to make the border of a submodel preview with round corners.
> 
> Are you saying that when you updated the submodel preview (to display round corners), the submodel was displayed on all the following steps - versus just the first step as expectred ?

No, if I activate submodel preview, the preview is rendered on all pages.
See this recording
https://drive.google.com/file/d/1R-yDWAiXPLRdOo-bvt43oCr8tZIkd6d_/view?usp=drivesdk
at the beginning.

Code:
0 !LPUB SUBMODEL_DISPLAY PLACEMENT GLOBAL TOP LEFT PLI OUTSIDE
0 !LPUB SUBMODEL_DISPLAY MODEL_SCALE GLOBAL  0.8000
0 !LPUB SUBMODEL_DISPLAY SHOW GLOBAL TRUE

> > the weird rendering I mentioned earlier came up again...
> 
> Are you talking about the rendering with artifacts #518 ?

Yes. See th PLI of step 11 at 0:35" of this recording
 https://drive.google.com/file/d/1rsyjTiNsQWESK2wPuqNX3tHi1_HvEL-F/view?usp=drivesdk

**Comment by trevorsandy** [comment]
Ok - got it. Thank you .

> if you directly navigate to say page 10...

If you jump to page 10 from which page ? This information is useful as jumping from a submodel/callout/step group page etc... will trigger uses different routines

> Redrawing the whole file or redraw a certain page

Same as above, can you indicate from which page you attempted to redraw that produces the unexpected behaviour ?

Cheers,

**Comment by jacovandermolen** [comment]
> Ok - got it. Thank you .
> 
> > if you directly navigate to say page 10...
> 
> If you jump to page 10 from which page ? This information is useful as jumping from a submodel/callout/step group page etc... will trigger uses different routines

In this case, I went from page 6 to 1 and then choose to go to page 10.
See https://drive.google.com/file/d/1rsyjTiNsQWESK2wPuqNX3tHi1_HvEL-F/view from 0:42" and on

> > Redrawing the whole file or redraw a certain page
> 
> Same as above, can you indicate from which page you attempted to redraw that produces the unexpected behaviour ?

Redraw on either page gets things messed up.

I have been thinking: would you be interested in doing a meeting sometime to let you see what is happening? You can watch my screen as I go and maybe we can talk a bit live?

**Comment by trevorsandy** [comment]
> would you be interested in doing a meeting sometime to let you see what is happening? You can watch my screen as I go and maybe we can talk a bit live?

That's a good idea. 

I have enough information for the this cycle. For the next update, we can take a session to see what's happening on your system.

Cheers,

**Comment by jacovandermolen** [comment]
Great. What time zone are you in? Since you seem te be online now, I think somewhere around 11am CET could be a good time?

**Comment by trevorsandy** [comment]
I'm in CET - (France)

Cheers,

**Comment by jacovandermolen** [comment]
LOL. Fun fact: I thought you were American. Sorry... ;-)

**Comment by trevorsandy** [comment]
LOL. "Fun fact" There is at least one American living in France !

Cheers,

**Comment by trevorsandy** [comment]
Jaco, I've treated all the items you highlighted except #518 which I cannot reproduce. However, I've merged some updates to the Native renderer from Leonardo so it's worthwhile taking a look to see if the intermittent artefacts persist.

I'll check in the updates shortly.

Cheers,

**Comment by jacovandermolen** [comment]
Great! I’ll take a look this evening!

**Comment by trevorsandy** [comment]
Hi Jaco. FYI, in  [2517 (r85)](https://github.com/trevorsandy/lpub3d/releases/tag/continuous), I forced the parse routine to build every page between the current page and the goto page when jumping ahead. However, I didn't like so much this approach so, in  [2528 (r96)](https://github.com/trevorsandy/lpub3d/releases/tag/continuous), I added new logic to parse the build mods between pages so the goto page will have the model in the state up to that point in the model file. This approach is much faster. What is still to do (I'm working on it now) can be best described as:
- Jump from page 2 to page 12 (all is well and the model is rendered as expected)
- Return to page 3 (all is well agin)
- Navigate to page 4 where there is a a build modification (all is not well here, the build modification on page 4 will be set to Remove, from when we jumped to page 12)

Cheers,

**Comment by trevorsandy** [comment]
Hi Jaco. I just checked in the latest build mod updates. It would be better for you to test with the DevOps build r99.

Cheers,

**Comment by jacovandermolen** [comment]
I have done some testing with my 40252 set and all goes very well! Tomorrow I will do some more testing with other models.
The latest DEV is very stable and fast.

**Comment by trevorsandy** [comment]
Hi Jaco, Excellent. What build did you use ? 

I saw the latest DevOps build 2762 (r100) for Windows failed to publish to GitHub. You can get it however directly from the AppVeyor LPub3D-CI [artifects repository](https://ci.appveyor.com/project/trevorsandy/lpub3d-ci/builds/38938032/artifacts).

Cheers,

**Comment by jacovandermolen** [comment]
I tested so far with 2.4.2.57.2719.
I will do some more testing with other models too with 2528 (r96).

(If I click the artifacts repository I get "Build not found or access denied.")

**Comment by trevorsandy** [comment]
I deleted the artefacts repo while resolving the failure to publish to GitHub. Anyway, I’ve since released [2533 (r101)](https://github.com/trevorsandy/lpub3d/releases/tag/continuous).

If there are no returns over the next few days, I’ll promote this release to 2.4.3.

Cheers,

**Comment by jacovandermolen** [comment]
Hi,
Sorry for my late reply, but RL got in the way of LEGO...
So far, everything seems to work. I need to pick up on some instructions and the manual and let you know if anything else comes up.

**Comment by trevorsandy** [comment]
Hi Jaco. No worries. Thanks for the update.

Cheers,

**Comment by jacovandermolen** [comment]
I have finished conducting some extensive testing with LEGO set 40252 and tried to come as close to the original instructions as I could. We are getting very close to being able to do that. There are a few minor things I will post on.
Here is the file
[40252LPub243x.mpd.txt](https://github.com/trevorsandy/lpub3d/files/6553420/40252LPub243x.mpd.txt)

**Comment by trevorsandy** [comment]
> Testing Buildmods now with 2.4.4.4.2675_20220612. Seems better, but still some hickups and bugs here and there.
We can perhaps look into this again?

This behaviour has been updated and will be available in release v2.4.5.

Specifically:

- I've ensured every page scenario in your model file 40252LPub243x.mpd is behaving as expected.
- fixed the randomly inconsistent page count behaviour

A **known issue** sometimes display an old image in the instruction page (showing the build mod) while the correct image is currently rendered in the visual editor. In this case, you can simply right-click (enable the context menu) on the current page and select Reset Assembly Image Cache.

![Screenshot - 03_07_2022 , 10_19_54](https://user-images.githubusercontent.com/13388970/177031383-875e260e-ce06-4ee6-824b-382578281934.png)

You can take an advanced look with the latest [DevOps build](https://github.com/trevorsandy/lpub3d-ci/releases/tag/continuous).

Let me know if your checks continue to produce unexpected behaviour. 

Cheers,
